worker/src/RTC/ActiveSpeakerObserver.cpp

```diff
@@ -0,0 +1,540 @@
+
+		this->interval = jsonIntervalIt->get<int16_t>();
+
+		if (this->interval < 100)
+			this->interval = 100;
+		else if (this->interval > 5000)
+			this->interval = 5000;
+
+		this->periodicTimer = new Timer(this);
+
+		this->periodicTimer->Start(interval, interval);
+	}
+
+	ActiveSpeakerObserver::~ActiveSpeakerObserver()
+	{
+		MS_TRACE();
+
+		delete this->periodicTimer;
+	}
+
+	void ActiveSpeakerObserver::AddProducer(RTC::Producer* producer)
+	{
+		MS_TRACE();
+
+		if (producer->GetKind() != RTC::Media::Kind::AUDIO)
+			MS_THROW_TYPE_ERROR("not an audio Producer");
+
+		this->mapProducerSpeaker[producer->id].producer = producer;
+		this->mapProducerSpeaker[producer->id].speaker  = new Speaker();
+	}
+
+	void ActiveSpeakerObserver::RemoveProducer(RTC::Producer* producer)
+	{
+		MS_TRACE();
+
+		if (mapProducerSpeaker[producer->id].speaker != nullptr)
+		{
+			delete mapProducerSpeaker[producer->id].speaker;
+		}
+
+		this->mapProducerSpeaker.erase(producer->id);
+
+		if (producer->id == this->dominantId)
+		{
+			Update();
+		}
+	}
+
+	void ActiveSpeakerObserver::ProducerResumed(RTC::Producer* producer)
+	{
+		MS_TRACE();
+
+		auto it = this->mapProducerSpeaker.find(producer->id);
+
+		if (it != this->mapProducerSpeaker.end())
+		{
+			auto& rtpObserver = it->second;
+
+			rtpObserver.speaker->paused = false;
+		}
+	}
+
+	void ActiveSpeakerObserver::ProducerPaused(RTC::Producer* producer)
+	{
+		MS_TRACE();
+
+		auto it = this->mapProducerSpeaker.find(producer->id);
+
+		if (it != this->mapProducerSpeaker.end())
+		{
+			auto& rtpObserver = it->second;
+
+			rtpObserver.speaker->paused = true;
+		}
+	}
+
+	void ActiveSpeakerObserver::ReceiveRtpPacket(RTC::Producer* producer, RTC::RtpPacket* packet)
+	{
+		MS_TRACE();
+
+		if (IsPaused())
+			return;
+
+		uint8_t volume;
+		bool voice;
+
+		if (!packet->ReadSsrcAudioLevel(volume, voice))
+			return;
+
+		auto it = this->mapProducerSpeaker.find(producer->id);
+
+		if (it != this->mapProducerSpeaker.end())
+		{
+			auto& rtpObserver = it->second;
+			uint64_t now      = DepLibUV::GetTimeMs();
+
+			rtpObserver.speaker->LevelChanged(volume, now);
+		}
+	}
+
+	void ActiveSpeakerObserver::Paused()
+	{
+		MS_TRACE();
+
+		this->periodicTimer->Stop();
+	}
+
+	void ActiveSpeakerObserver::Resumed()
+	{
+		MS_TRACE();
+
+		this->periodicTimer->Restart();
+	}
+
+	void ActiveSpeakerObserver::OnTimer(Timer* timer)
+	{
+		MS_TRACE();
+
+		Update();
+	}
+
+	void ActiveSpeakerObserver::Update()
+	{
+		MS_TRACE();
+
+		uint64_t now              = DepLibUV::GetTimeMs();
+		uint64_t levelIdleTimeout = LevelIdleTimeout - (now - lastLevelIdleTime);
+		if (levelIdleTimeout <= 0)
+		{
+			if (lastLevelIdleTime != 0)
+			{
+				TimeoutIdleLevels(now);
+			}
+			lastLevelIdleTime = now;
+		}
+
+		if (mapProducerSpeaker.size() > 0 && CalculateActiveSpeaker())
+		{
+			json data          = json::object();
+			data["producerId"] = this->dominantId;
+			Channel::ChannelNotifier::Emit(this->id, "dominantSpeaker", data);
```

Comment:
Rename `dominantSpeaker` to `dominantspeaker`. In fact this was already changed in JS side.

```suggestion
			data["producerId"] = this->dominantId;

			Channel::ChannelNotifier::Emit(this->id, "dominantspeaker", data);
```